libraries/common/scanner.effekt

```diff
@@ -0,0 +1,273 @@
+  val t = do peek[A]()
+  if (predicate(t)) {
+    do skip[A]()
+    return t
+  } else {
+    do fail()
+  }
+}
+
+/// Check that the next token satisfies the predicate and skip it when it does.
+def skipIf[A] { predicate: A => Bool }: Unit / { Scan[A], fail, stop } = {
+  readIf{predicate}; ()
+}
+
+/// Reads until the predicate does not hold for the next token.
+/// Emits the tokens read.
+def readWhile[A] { predicate: A => Bool }: Unit / { Scan[A], emit[A], stop } =
+  successfully { do emit(readIf{predicate}) }
+
+/// Skips until the predicate does not hold for the next token.
+def skipWhile[A] { predicate: A => Bool }: Unit / { Scan[A], stop } =
+  for[A] { readWhile{predicate} } { _ => () }
+
+
+/// Read the next character if it is the given one and return true in this case.
+def readIf(e: Char): Unit / { Scan[Char], fail, stop } = {
+  readIf[Char] { c => c == e}; ()
+}
+
+/// Check that the next token satisfies the predicate and skip it when it does.
+def skipIf(e: Char): Unit / { Scan[Char], fail, stop } =
+  skipIf { c => c == e }
+
+/// Skip until the next character is not a whitespace character
+def skipWhitespace(): Unit / { Scan[Char], stop } =
+  skipWhile { c => c.isWhitespace }
+
+
+def prettyUnexpectedChar(got: Char, expected: Char): String =
+  "Unexpected character, got '" ++ got.toString ++ "', but expected '" ++ expected.toString ++ "'"
+
+def unexpectedChar(got: Char, expected: Char): Nothing / Exception[WrongFormat] =
+  wrongFormat(prettyUnexpectedChar(got, expected))
+
+/// Expect a character class with a name and raise an exception if it fails or stops.
+def expect[R](name: String) { scanner: () => R / {fail, stop} }: R / { Scan[Char], Exception[WrongFormat] } =
+  try {
+    try {
+      scanner()
+    } with fail {
+      wrongFormat("Expected " ++ name ++ " but got " ++ do peek[Char]().toString)
+    }
+  } with stop {
+    wrongFormat("Unexpected end of input")
+  }
+
+/// Require that the next character is the given one, raising an exception otherwise.
+def expect(c: Char): Unit / { Scan[Char], Exception[WrongFormat] } =
+  expect(c.toString) { readIf(c); () }
+
+/// Check that the next characters are the given String, raising an exception otherwise.
+/// Advances over the longest common prefix with this String.
+def expect(s: String): Unit / { Scan[Char], Exception[WrongFormat] } =
+  for[Char] { s.each } { c => expect(c); () }
+
+/// Check that the next character is a digit in base 10, and if so read and return it.
+def readDigit(): Int / { Scan[Char], fail, stop } =
+  if (digitValue(do peek[Char]) is Some(d)) {
+    do skip[Char]()
+    return d
+  } else {
+    do fail()
+  }
+
+/// Read a positive decimal number.
+def readDecimal(): Int / { Scan[Char], stop, Exception[WrongFormat] } = {
+  var result = expect("digit") { readDigit() }
+  for[Int] { successfully { readDigit() } } { digit =>
+    result = result * 10 + digit
+  }
+  result
+}
```

Comment:
This ^ doesn't really work, the `successfully` eats up everything it sees, so `readDecimal` always returns only the first digit of the parsed number.
Instead, I'd think you want something more like:
```suggestion
/// Read a positive decimal number.
def readDecimal(): Int / { Scan[Char], stop, Exception[WrongFormat] } = {
  var result = expect("digit") { readDigit() }
  successfully { 
    result = result * 10 + readDigit()
  }
  result
}
```